### PR TITLE
Improving the raytraced reflection quality modes

### PR DESCRIPTION
**Pull request by anisunity**
Implementing a custom low budget filter for quarter resolution reflection calculus
Adding an integration mode for reflections
Using Temporal Anti-aliasing to filter reflections
Correcting a bug in the max taa frame
Adding an optional raytraced reflection history texture
Using O2 SAmpling for the reflection effect

### CODE REVIEW

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingReflectionFilter.compute [hunk 1]
```diff
@@ -0,0 +1,239 @@
+#pragma kernel RaytracingReflectionFilter
+#pragma kernel TemporalAccumulationFilter
+
+#include "Packages/com.unity.render-pipelines.core/ShaderLibrary/Common.hlsl"
+#include "Packages/com.unity.render-pipelines.high-definition\Runtime\ShaderLibrary\ShaderVariables.hlsl"
+#include "Packages/com.unity.render-pipelines.high-definition/Runtime/Material/NormalBuffer.hlsl"
+#include "Packages/com.unity.render-pipelines.core/ShaderLibrary/BSDF.hlsl"
+#include "Packages/com.unity.render-pipelines.high-definition\Runtime\Material\PreIntegratedFGD\PreIntegratedFGD.hlsl"
```
Comment (Chman) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingReflectionFilter.compute]: Mixed slashes and backslashes, let's be safe and use Unix paths everywhere.

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Camera/HDCamera.cs [hunk 2]
```diff
@@ -617,7 +622,7 @@ Matrix4x4 GetJitteredProjectionMatrix(Matrix4x4 origProj)
             float jitterY = HaltonSequence.Get((taaFrameIndex & 1023) + 1, 3) - 0.5f;
             taaJitter = new Vector4(jitterX, jitterY, jitterX / camera.pixelWidth, jitterY / camera.pixelHeight);
 
-            const int kMaxSampleCount = 256;
+            const int kMaxSampleCount = 8;
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Camera/HDCamera.cs]: have you discuss this with Thomas ?

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/OnlineVariance.hlsl [hunk 3]
```diff
@@ -0,0 +1,43 @@
+struct VarianceEstimator
+{
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/OnlineVariance.hlsl]: why is this call OnlineVariance.hlsl ?

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingConsts.hlsl [hunk 4]
```diff
@@ -0,0 +1,2 @@
+// This array converts an index to the local coordinate shift of the half resolution texture
+static const uint2 HalfResIndexToCoordinateShift[4] = { uint2(0,0), uint2(1, 0), uint2(0, 1), uint2(1, 1) };
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingConsts.hlsl]: caution about const here (let's talk)

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingReflections.raytrace [hunk 5]
```diff
@@ -10,122 +10,251 @@
...
     
-    // Check if this pixel should not recieve ssr, for the moment we compute it for everyone
-    bool doesntReceiveSSR = (stencilVal & _SsrStencilExclusionValue) != 0;
-    if (doesntReceiveSSR)
-    {
+    // Decode the world space normal
+    NormalData normalData;    
+    DecodeFromNormalBuffer(sourceCoord, normalData);
+
+    // If this value is beyond the smothness that we allow, no need to compute it
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingReflections.raytrace]: smoothness

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingReflections.raytrace [hunk 6]
```diff
@@ -10,122 +10,251 @@
...
+    uint2 currentCoord = uint2(LaunchIndex.x, LaunchDim.y - LaunchIndex.y - 1);
+
+    // Clear the output color texture
+    _SsrLightingTextureRW[currentCoord] = float4(0.0f, 0.0f, 0.0f, 0.0f);
+
+    // Read the depth value
+    float depthValue  = _DepthTexture[currentCoord];
+
+    // This point is part of the background, we don't really care 
+    if(depthValue == 0.0f)
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingReflections.raytrace]: Do a pass on the code and check all part where you do this test, you must not assume it is 0.0 but use UNITY_RAW_FAR_CLIP_VALUE

            if (depth == UNITY_RAW_FAR_CLIP_VALUE)
            {
                // When a pixel is at far plane, the world space coordinate reconstruction is not reliable.
                // So in order to have a valid position (for example for height fog) we just consider that the sky is a sphere centered on camera with a radius of 5km (arbitrarily chosen value!)
                // And recompute the position on the sphere with the current camera direction.
                posInput.positionWS = GetCurrentViewPosition() - V * _MaxFogDistance;

                // Warning: we do not modify depth values. Do not use them!
            }

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingReflections.raytrace [hunk 7]
```diff
@@ -10,122 +10,251 @@
...
+    float roughness = PerceptualRoughnessToRoughness(normalData.perceptualRoughness);
+
+    // Compute the scramble for the current pixel (must tbe the same across pixels)
+    const uint2 scramble = GetRaytracingNoiseSample(currentCoord, 0) * 0xffffffff;
+
+    // Variable that accumulate the radiance
+    float3 finalColor = float3(0.0, 0.0, 0.0);
+
+    // Loop through the samples and add their contribution
+    for(int sampleIndex = 0; sampleIndex < _RaytracingNumSamples; ++sampleIndex)
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingReflections.raytrace]: for (
if (
don't miss the space

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingSampling.hlsl [hunk 8]
```diff
@@ -3,6 +3,33 @@ Texture2DArray<float4>						_RaytracingNoiseTexture;
 uint 										_RaytracingNoiseResolution;
 uint 										_RaytracingNumNoiseLayers;
 
+float GetRadicalInverse(uint n, uint scramble)
+{
+    n = (n << 16) | (n >> 16);
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/RaytracingSampling.hlsl]: please use uint ReverseBits32(uint bits) (see Hammersley)

isn't this: 

real VanDerCorputBase2(uint i)
{
    return ReverseBits32(i) * rcp(4294967296.0); // 2^-32
}

Let's try to have a single file with all this sampling functions insteal of spreading them at different palce

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/ShaderVariablesRaytracing.hlsl [hunk 9]
```diff
@@ -0,0 +1,11 @@
+// The target acceleration acceleration structure
+RaytracingAccelerationStructure         _RaytracingAccelerationStructure;
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/ShaderVariablesRaytracing.hlsl]: why is this needed in Lit.shader ? and if it is the case it must be guardbanded for raytracing define

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/OnlineVariance.hlsl (hunk 3 above)
Comment (anisunity) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/OnlineVariance.hlsl]: https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford's_Online_algorithm

Comment (anisunity) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/OnlineVariance.hlsl]: (i'll add the link to the Code)
